libraries/AsyncUDP/src/AsyncUDP.cpp

```diff
@@ -324,6 +324,22 @@ AsyncUDPPacket::AsyncUDPPacket(AsyncUDP *udp, pbuf *pb, const ip_addr_t *raddr,
     }
 }
 
+AsyncUDPPacket::AsyncUDPPacket(const AsyncUDPPacket &packet)
+    : _udp(packet._udp),
+      _pb(packet._pb),
+      _if(packet._if),
+      _localIp(packet._localIp),
+      _localPort(packet._localPort),
+      _remoteIp(packet._remoteIp),
+      _remotePort(packet._remotePort),
+      _data(packet._data),
+      _len(packet._len),
+      _index(packet._index)
+{
+    memcpy(_remoteMac, packet._remoteMac, sizeof(_remoteMac));
+    pbuf_ref(_pb);
+}
+
 AsyncUDPPacket::~AsyncUDPPacket()
 {
     pbuf_free(_pb);
```

Comment:
here the pbuf should be freed. should take care of `pbuf_ref` as on that depends if `pbuf_free` will work

Reply:
It is freed, as you point out, in the destructor. This balances the call to `pbuf_ref` in the constructor. That's why my change also adds a call to `pbuf_ref` in the copy constructor. This solves the imbalance.

Reply:
As written, the code won't successfully work for _both_ pass-as-reference and pass-as-copy forms, no matter where the `pbuf_free` call is used in `AsyncUDP::_recv`, either outside or inside an `else`. That's why adding to the copy constructor solves the problem for both cases, if the `pbuf_free` is moved outside the if/else.

Reply:
I have tried your fix before and it did not work in many situations. so I reverted it back. you could look in the history... should be there :) need to figure out better that pbuf ref/unref count and not leak or double free

Reply:
Interesting. Could you show me some situations in which my fix fails? I’d like to see if I can resolve them. All I know is that my fix doesn’t fail in any of my tests.

Reply:
unfortunately I can not recall...

Reply:
So you’re saying that in the past you’ve already tried adding this fix _with_ changing the copy constructor and it didn’t work? What do you think of the logic in #3288? I don’t see what’s wrong with it.

Could you at least show which commits you undid where the fix didn’t work?

Reply:
@me-no-dev ?